lib/afinter.c

```diff
@@ -635,3 +644,141 @@ afinter_get_metrics(void)
 {
   return metrics;
 }
+
+/****************************************************************************
+ * Live collection of internal log messages.
+ ****************************************************************************/
+
+AFInterLive
+afinter_start_live_collection(void)
+{
+  g_mutex_lock(&internal_msg_lock);
+
+  if (is_live_collection)
+    {
+      g_mutex_unlock(&internal_msg_lock);
+      return AFINTER_LIVE_COLLECTION_RUNNING;
+    }
+
+  if (current_internal_source != NULL)
+    {
+      g_mutex_unlock(&internal_msg_lock);
+      return AFINTER_INTERNAL_SRC_PRESENT;
+    }
+
+  AFInterSource *self = g_new0(AFInterSource, 1);
+  AFInterSourceOptions *options = g_new0(AFInterSourceOptions, 1);
+
+  afinter_source_options_defaults(options);
+  self->options = options;
+  stats_counter_set(metrics.queue_capacity, self->options->queue_capacity);
+
+  current_internal_source = self;
+  is_live_collection = TRUE;
+
+  g_mutex_unlock(&internal_msg_lock);
+
+  return AFINTER_LIVE_COLLECTION_STARTED;
+}
+
+static void
+afinter_reset_live_collection(void)
+{
+  afinter_stop_live_collection(NULL);
+
+  g_mutex_lock(&internal_msg_lock);
+  if (is_live_collection)
+    {
+      if (internal_msg_queue)
+        _release_internal_msg_queue();
+    }
+  g_mutex_unlock(&internal_msg_lock);
+}
+
+static void
+afinter_get_collected_messages(GString *result)
+{
+  LogMessage *log_msg;
+  GString *msg;
+  LogTemplate *template = NULL;
+
+  if (result == NULL)
+    return;
+
+  g_mutex_lock(&internal_msg_lock);
+  if (internal_msg_queue != NULL)
+    {
+      msg = g_string_sized_new(1024);
+
+      while (!g_queue_is_empty(internal_msg_queue))
+        {
+          log_msg = g_queue_pop_head(internal_msg_queue);
+          /* format log */
+          template = log_template_new(configuration, NULL);
+          log_template_compile(template, "[$DATE] $HOST $MSGHDR$MSG", NULL);
+          log_template_format(template, log_msg, &DEFAULT_TEMPLATE_EVAL_OPTIONS, msg);
+          g_string_append_printf(result, "%s\n", msg->str);
+          log_msg_unref(log_msg);
+          log_msg = NULL;
+          stats_counter_dec(internal_queue_length);
+        }
+
+      g_string_free(msg, TRUE);
+      internal_msg_queue = NULL;
+    }
```

Comment:
i can't see why this would be needed, as the queue itself should/can be reused in case of a future start/stop call

also it lacks any kind of free of the queue (either via _release_internal_msg_queue or g_queue_free_full, etc.)

Reply:
ok, but why do you release the queue? 

though it is reallocated in the `afinter_message_posted(LogMessage *msg)` at the next message posted, but in that case the counters will be re-initialized as well (that is surely wrong, as there was no reset of them anywhere)

you emptied the queue, but you can leave the container itself as it is

`afinter_global_deinit` will take care of it when it is not needed anymore, meaningless to destroy the container as I can see

Reply:
@HofiOne : so, should I remove `g_queue_free_full ` ?

Reply:
@HofiOne : okay removed the `free` call